libbpf-tools/memleak.c

```diff
@@ -0,0 +1,707 @@
+	.count = 0,
+	.show_allocs = false,
+	.min_age_ns = 500 * 1e6,
+	.command = NULL,
+	.combined_only = false,
+	.wa_missing_free = false,
+	.sample_rate = 1,
+	.top = 10,
+	.min_size = 0,
+	.max_size = 0,
+	.obj = NULL,
+	.percpu = false,
+	.kernel_trace = false,
+	.max_stack_depth = 127,
+};
+
+const char *argp_program_version = "memleak 0.1";
+const char *argp_program_bug_address = "https://github.com/iovisor/bcc/tree/master/libbpf-tools";
+const char argp_program_doc[] =
+"Trace and display outstanding allocations to detect\n"
+"memory leaks in user-mode processes and the kernel.\n"
+"\n"
+"USAGE: memleak [-h] [-p PID] [-t] [-a] [-o OLDER] [-c COMMAND]\n"
+"                [--combined-only] [--wa-missing-free] [-s SAMPLE_RATE]\n"
+"                [-T TOP] [-z MIN_SIZE] [-Z MAX_SIZE] [-O OBJ]\n"
+"                [interval] [count]\n"
+"\n"
+"EXAMPLES:\n"
+"\n"
+"./memleak -p $(pidof allocs)\n"
+"        Trace allocations and display a summary of \"leaked\" (outstanding)\n"
+"        allocations every 5 seconds\n"
+"./memleak -p $(pidof allocs) -t\n"
+"        Trace allocations and display each individual allocator function call\n"
+"./memleak -ap $(pidof allocs) 10\n"
+"        Trace allocations and display allocated addresses, sizes, and stacks\n"
+"        every 10 seconds for outstanding allocations\n"
+"./memleak -c \"./allocs\"\n"
+"        Run the specified command and trace its allocations\n"
+"./memleak\n"
+"        Trace allocations in kernel mode and display a summary of outstanding\n"
+"        allocations every 5 seconds\n"
+"./memleak -o 60000\n"
+"        Trace allocations in kernel mode and display a summary of outstanding\n"
+"        allocations that are at least one minute (60 seconds) old\n"
+"./memleak -s 5\n"
+"        Trace roughly every 5th allocation, to reduce overhead\n";
+
+static const struct argp_option opts[] = {
+	{ "pid", 'p', "PID", 0, "Process PID" },
+	{ "trace", 't', NULL, 0, "Print trace message for each alloc/free call" },
+	{ "show-allocs", 'a', NULL, 0, "Show allocation addresses and sizes as well as call stacks" },
+	{ "older", 'o', "MILLISECONDS", 0, "Prune allocations younger than this age in milliseconds" },
+	{ "command", 'c', "COMMAND", 0, "Execute and trace the specified command" },
+	{ "combined-only", 'C', NULL, 0, "Show combined allocation statistics only" },
+	{ "wa-missing-free", 'W', NULL, 0, "Workaround to alleviate misjudgments when free is missing" },
+	{ "sample-rate", 's', "RATE", 0, "Sample every N-th allocation to decrease the overhead" },
+	{ "top", 'T', NULL, 0, "Display only this many top allocating stacks (by size)" },
+	{ "min-size", 'z', NULL, 0, "Capture only allocations larger than this size" },
+	{ "max-size", 'Z', NULL, 0, "Capture only allocations smaller than this size" },
+	{ "obj", 'O', "OBJ", 0, "Attach to allocator functions in the specified object" },
+	{ "percpu", 'P', NULL, 0, "trace percpu allocations" },
+	{ NULL, 'h', NULL, OPTION_HIDDEN, "Show the full help" },
```

Comment:
Please add a `-v` option like other tools.

Reply:
Fix in commit [2beb289e](https://github.com/iovisor/bcc/pull/3828/commits/2beb289e98d651bddc2127299c5627b5d6149fce).